### PR TITLE
18 Add search

### LINKED ISSUE #18: Separate search from filtering
Right now, django_bread just provides filtering and expects the user to implement search with a custom filter if they want it. But searching is a very basic feature. It might be better if we go ahead and build in some support for it, to make it very easy to add search to views.

### PR DESCRIPTION
**Pull request by dpoirier**
Fixes #18

### CODE REVIEW

File: bread/bread.py [hunk 1]
```diff
@@ -182,11 +187,26 @@ def get_queryset(self):
         if self.filterset is not None:
             self.filter = self.filterset(self.request.GET, queryset=qset)
             qset = self.filter.qs
+
+        # Now search
+        q = self.request.GET.get('q', False)
+        if self.search_fields and q:
+            # print("Applying search. search_fields=%r, q=%r" % (self.search_fields, q))
```
Comment (vkurup) [inline bread/bread.py]: Did you mean to remove these, or switch them to logging calls?

File: bread/bread.py [hunk 2]
```diff
@@ -201,6 +221,41 @@ def get_context_data(self, **kwargs):
...
+        use_distinct = False
+        search_fields = self.search_fields
+        if search_fields and search_term:
+            orm_lookups = [construct_search(str(search_field))
+                           for search_field in search_fields]
+            for bit in search_term.split():
+                or_queries = [Q(**{orm_lookup: bit})
+                              for orm_lookup in orm_lookups]
+                queryset = queryset.filter(reduce(or_, or_queries))
+            if not use_distinct:
```
Comment (vkurup) [inline bread/bread.py]: I don't see how `use_distinct` would be anything but False at this point. Is the check necessary?

File: bread/templates/bread/includes/browse.html [hunk 3]
```diff
@@ -9,28 +9,41 @@
...
-           {% for err in filter.form.non_field_errors %}
-             <div class="alert alert-error help-block">
-               {{ err }}
-             </div>
-           {% endfor %}
-         </div>
-       {% endif %}
+     {% if has_search %}
+       <div class="control-group">
+         <p>{% blocktrans with names=view.bread.verbose_name_plural %}Search {{ names }}{% endblocktrans %}</p>
```
Comment (vkurup) [inline bread/templates/bread/includes/browse.html]: `verbose_name_plural` should be in the template context, so this should be `names=verbose_name_plural`.

File: tests/test_search.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
+try:
+    # Python 2
+    from httplib import OK
+except ImportError:
+    # Python 3
+    from http.client import OK
```
Comment (vkurup) [inline tests/test_search.py]: could use `six` here.

File: tests/test_search.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        obj_ids = [obj.id for obj in objs]
+        self.assertEqual([self.joe.id], obj_ids)
+
+    def test_multiple_terms_match(self):
+        # A record that matches all terms is returned
+        objs = self.get_search_results(q='Joe Smith')
+        obj_ids = [obj.id for obj in objs]
+        self.assertEqual([self.joe.id], obj_ids)
+
+    def test_multiple_terms_dont_match(self):
```
Comment (vkurup) [inline tests/test_search.py]: Not sure if it's valuable, but maybe a test with a query which matches a different field in different records? In this case, `q=o` or `q=i`?

**Comment by vkurup** [comment]
A few minor comments, then :mag:

File: bread/bread.py (hunk 1 above)
Comment (dpoirier) [inline bread/bread.py]: On Thu, May 7, 2015 at 10:47 AM, Vinod Kurup <notifications@github.com>
wrote:

> Did you mean to remove these, or switch them to logging calls?


​Oops. Removed.​

File: bread/bread.py (hunk 2 above)
Comment (dpoirier) [inline bread/bread.py]: It doesn't look necessary, does it? I just copied this whole block of code from Django and made the minimum changes. It's probably not needed in the original code either, but it appears harmless, so I'd rather leave it in for now to keep as close as possible to the original code. That way if we ever need to pick up a bug fix to this code from Django, it'll be easier.

**Pull request merged**